README.md
# Block Chain RPC

Provide basic RPC functionalities for various Chain, in elixir way.

## Installation

If [available in Hex](https://hex.pm/docs/publish), the package can be installed
by adding `ocap_rpc` to your list of dependencies in `mix.exs`:

```elixir
def deps do
  [
    {:ocap_rpc, "~> 0.0.0"}
  ]
end
```

## Configuration
The following is the default configs, but you don't need to define all of them in your project.

```elixir
config :ocap_rpc, :eth,
  conn: %{
    hostname: "localhost",
    port: 8545
  },
  timeout: 5_000,
  chain_id: 1

config :ocap_rpc, :btc,
  conn: %{
    hostname: "localhost",
    port: 8332
  },
  timeout: 5_000

config :ocap_rpc, :ipfs,
  conn: %{
    hostname: "localhost",
    port: 5001
  },
  timeout: 5_000

config :ocap_rpc, :cmt,
  conn: %{
    hostname: "localhost",
    port: 8545
  },
  timeout: 5_000
```

If you just want to use one of these chain, you can do something like:

```elixir
config :ocap_rpc, :eth, chain_id: 31337
```

## Environment Variable Dependencies

To let ocap_rpc work correctly, please make sure you have following environment variables set in runtime environment.

  - BTC_RPC_HOST
  - BTC_RPC_USER
  - BTC_RPC_PASS
  - ETH_RPC_HOST
  - CMT_RPC_HOST

Documentation can be generated with [ExDoc](https://github.com/elixir-lang/ex_doc)
and published on [HexDocs](https://hexdocs.pm). Once published, the docs can
be found at [https://hexdocs.pm/ex_bitcoin](https://hexdocs.pm/ex_bitcoin).


CHANGELOG.md
## 0.15.3 (Jun 10, 2022)
  - Allow users not to put configs for all chains in their config.exs

## 0.15.2 (December 30, 2019)
  - Make sure the RPC return nil if the tx is not persisted on chain.

## 0.15.1 (December 20, 2019)
  - Fix nonce issue.

## 0.15.0 (December 20, 2019)
  - Add APIs to return unsigned transaction and ERC20 transfer.

## 0.14.5 (October 10, 2019)
  - Add compose transaction and transfer

## 0.14.4 (August 21, 2019)
  - Update deps version

## 0.14.3 (August 20, 2019)
  - Supports sending ethereum transaction.

## 0.14.2 (August 15, 2019)
  - fix tests
  - Fix the problem caused by ethereum/abi not being able to decode input data with extra bytes.

## 0.14.1 (November 29, 2018)
  - delete distillery and make abi optional

## 0.14.0 (November 28, 2018)
  - Bring  ocap rpc to use elixir 1.7.4
  - support cmt basic rpc (#39)

## 0.13.4 (November 13, 2018)
  - Handle the case where the block returned is null.
  - Support cmt basic rpc.

## 0.13.3 (November 06, 2018)
  - Disable the unnecessary logs.

## 0.13.2 (November 02, 2018)
  - Updates for arc block listener ethereum v2.

## 0.13.1 (November 01, 2018)
  - update args in ipfs rpc

## 0.13.0 (October 31, 2018)
  - support most of the ipfs APIs (note that the file upload is not fully functional yet)
  - support the first ipfs rpc

## 0.12.8 (October 30, 2018)
  - Add logic to handle the case where no traces retrieved for calculating block reward.

## 0.12.7 (October 28, 2018)
  - Use the new abi to handle the memory leak.

## 0.12.6 (October 27, 2018)
  - Resolve the ignored transactions at runtime.

## 0.12.5 (October 27, 2018)
  - Ignore transactions that could cause memory leak while decoding ABI.

## 0.12.4 (October 27, 2018)
  - Stop decoding huge transaction input to prevent potential memory leak.

## 0.12.3 (October 25, 2018)
  - Update the logic to handle unprintable string.

## 0.12.2 (October 25, 2018)
  - Update the logic to parse error and receipt status in trace.

## 0.12.1 (October 24, 2018)
  - Add time and receipt status to trace.

## 0.12.0 (August 28, 2018)
  - fix transaction tx type issue and parse contract from / to better
  - retrieve block reward and change unit to wei

## 0.11.8 (September 13, 2018)
  - Bypass the contract_to parsing issue for now.

## 0.11.7 (September 12, 2018)
  - Handle the case where block number is out of range.

## 0.11.6 (August 31, 2018)
 - Disable transaction trace for now.

## 0.11.5 (August 31, 2018)
  - Add code to handle error.

## 0.11.4 (August 29, 2018)
 - move timeout value to config

## 0.11.3 (August 25, 2018)
  - add tx time

## 0.11.1 (August 24, 2018)
  - Change namespace of PoisonedDecimal.

## 0.11.0 (August 24, 2018)
  - support basic btc rpc

## 0.10.3 (August 23, 2018)
  - Remove 0x prefix for pre_hash

## 0.10.2 (August 23, 2018)
  - fix various issues and use wei in all values

## 0.10.1 (August 21, 2018)
  - Fix the readme.md

## 0.10.0 (August 16, 2018)
  - replace httpoison with tesla
  - add retry support

## 0.9.7 (August 10, 2018)
  - fix sentry issue


## 0.9.0 (August 01, 2018)
  - allow fees and reward calculation.

## 0.8.0 (August 01, 2018)
  - refactor the data type
  - refactor the yaml files

## 0.7.2 (July 30, 2018)


## 0.7.0 (July 29, 2018)
  - support decode to erc20

## 0.6.0 (July 29, 2018)
  - add basic test suites

## 0.5.0 (July 28, 2018)
  - update the code to support more functionalities

## 0.4.0 (July 27, 2018)
  - rename blockchain rpc to ocap rpc

## 0.3.0 (July 27, 2018)
  - support contract transactions

## 0.2.1 (July 27, 2018)


## 0.2.0 (July 26, 2018)


mix.exs
defmodule OcapRpc.MixProject do
  use Mix.Project

  @version File.cwd!() |> Path.join("version") |> File.read!() |> String.trim()
  @elixir_version File.cwd!() |> Path.join(".elixir_version") |> File.read!() |> String.trim()
  @otp_version File.cwd!() |> Path.join(".otp_version") |> File.read!() |> String.trim()

  def get_version, do: @version
  def get_elixir_version, do: @elixir_version
  def get_otp_version, do: @otp_version

  def project do
    [
      app: :ocap_rpc,
      version: @version,
      elixir: @elixir_version,
      elixirc_paths: elixirc_paths(Mix.env()),
      compilers: Mix.compilers(),
      start_permanent: Mix.env() == :prod,
      deps: deps()
    ]
  end

  # Run "mix help compile.app" to learn about applications.
  def application do
    [
      extra_applications: [:logger],
      mod: {OcapRpc.Application, []}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(:dev), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  # Run "mix help deps" to learn about dependencies.
  defp deps do
    [
      # convert
      {:atomic_map, "~> 0.9"},
      {:decimal, "~> 1.6"},
      {:hexate, ">= 0.6.0"},
      {:proper_case, "~> 1.3"},
      {:recase, "~> 0.6"},
      {:yaml_elixir, "~> 2.1"},

      # http client
      {:tesla, "~> 1.2"},
      {:jason, "~> 1.1"},

      # Ethereum related deps
      {:abi, git: "https://github.com/arcblock/abi.git"},
      # {:abi, path: "/Users/peiling/Documents/GitHub/ArcBlock/abi"},
      {:exth_crypto, "~> 0.1"},
      {:ex_rlp, "~> 0.5.1"},

      # utility belt
      {:utility_belt, "~> 0.16"},

      # dev & test
      {:credo, "~> 1.0", only: [:dev, :test], runtime: false},
      {:dialyxir, "~> 0.5", only: [:dev], runtime: false},
      {:ex_doc, "~> 0.19", only: [:dev, :test], runtime: false},
      {:excheck, "~> 0.6", only: :test, runtime: false},
      {:pre_commit_hook, "~> 1.2", only: [:dev, :test], runtime: false},
      {:triq, "~> 1.3", only: :test, runtime: false},

      # test only
      {:ex_machina, "~> 2.2", only: [:dev, :test]},
      {:faker, "~> 0.11", only: [:dev, :test]},
      {:mock, "~> 0.3", only: [:dev, :test]}
    ]
  end
end


lib/ipfs/rpc.ex
defmodule OcapRpc.Internal.IpfsRpc do
  @moduledoc """
  RPC request to ipfs server
  """
  use Tesla
  require Logger
  alias OcapRpc.Internal.Utils

  @version "v0"

  # TODO(lei): when tesla not compatible issue solved: `https://github.com/teamon/tesla/issues/157`
  if Application.get_env(:ocap_rpc, :env) not in [:test] do
    plug(Tesla.Middleware.Timeout, timeout: Application.get_env(:ocap_rpc, :timeout, 240_000))
  end

  def call(method, verb, args) do
    %{hostname: hostname, port: port} = Utils.get_connection(:ipfs)

    url = "http://#{hostname}:#{to_string(port)}/api/#{@version}/#{method}"

    Logger.debug(fn ->
      "IPFS RPC request to #{verb} for: #{inspect(url)}. args: #{inspect(args)}"
    end)

    result = apply(Tesla, verb, [url, [query: args]])

    case result do
      {:ok, %{status: 200, body: body, headers: headers}} ->
        content_type = get_header(headers, "content-type")

        case content_type do
          "application/json" -> Jason.decode!(body)
          _ -> body
        end

      {:ok, %{status: status, body: body}} ->
        Logger.error("Status: #{status}. Result: #{inspect(body)}")
        raise RuntimeError

      {:error, reason} ->
        raise(
          "RPC call failed. Reason: #{inspect(reason)}, method: #{inspect(method)}, arguments: #{
            inspect(args)
          }"
        )

      e ->
        Logger.error(e)
        raise RuntimeError
    end
  end

  defp get_header(headers, name) do
    Enum.reduce_while(headers, nil, fn {k, v}, _acc ->
      if k === name, do: {:halt, v}, else: {:cont, nil}
    end)
  end
end


lib/eth/advance/helper.ex
defmodule OcapRpc.Internal.EthTransaction.Helper do
  @moduledoc """
  Contains helper methods to construct and sign Ethereum transactions.
  """

  alias ExthCrypto.Hash.Keccak
  alias ExthCrypto.Signature
  alias OcapRpc.Internal.Utils

  @chain_ids %{
    ethereum_mainnet: 1,
    morden: 2,
    expanse_mainnet: 2,
    ropsten: 3,
    rinkeby: 4,
    ubiq_mainnet: 8,
    ubiq_testnet: 9,
    rootstock_mainnet: 30,
    rootstock_testnet: 31,
    kovan: 42,
    ethereum_classic_mainnet: 61,
    ethereum_classic_testnet: 62,
    ewasm_testnet: 66,
    geth_private_chains: 1337,
    görli: 6284,
    stureby: 314_158
  }

  @doc """
  Generates the signature of the `transaction_digest` by using the `private_key`. The transaction digest
  is the kecacak256 hash of the RLP encoded list of nine elements:
  [nonce, gasprice, gaslimit, to, value, data, v, r, s]

  Returns `{signature, r, s, v}`
  """
  @spec get_signature(binary(), String.t()) :: {binary(), integer(), integer(), integer()}
  def get_signature(transaction_digest, private_key) do
    Signature.sign_digest(
      transaction_digest,
      Utils.hex_to_binary(private_key)
    )
  end

  @doc """
  Gets the digest of the transaction to sign.
  """
  @spec get_transaction_digest_to_sign(
          integer(),
          integer(),
          integer(),
          String.t(),
          integer(),
          String.t()
        ) :: binary()
  def get_transaction_digest_to_sign(nonce, gas_price, gas_limit, to, value, input) do
    nonce
    |> get_transaction_to_sign(gas_price, gas_limit, to, value, input)
    |> Keccak.kec()
  end

  @doc """
  Gets the transaction to sign.
  """
  @spec get_transaction_to_sign(
          integer(),
          integer(),
          integer(),
          String.t(),
          integer(),
          String.t()
        ) :: binary()
  def get_transaction_to_sign(nonce, gas_price, gas_limit, to, value, input) do
    # EIP155 spec:
    # when computing the hash of a transaction for purposes of signing or recovering,
    # instead of hashing only the first six elements (ie. nonce, gasprice, startgas, to, value, data),
    # hash nine elements, with v replaced by CHAIN_ID, r = 0 and s = 0
    chain_id = Utils.get_chain_id() |> IO.inspect(label: "chain_id")

    nonce
    |> new_transaction(gas_price, gas_limit, to, value, input, chain_id, 0, 0)
    |> ExRLP.encode()
  end

  @doc """
  Gets the raw transaction which is ready to broadcast to the network.
  """
  @spec get_raw_transaction(
          integer(),
          integer(),
          integer(),
          String.t(),
          integer(),
          String.t(),
          integer(),
          integer(),
          integer()
        ) :: String.t()
  def get_raw_transaction(nonce, gas_price, gas_limit, to, value, input, v, r, s) do
    recovery_id = get_recovery_id(v)

    nonce
    |> new_transaction(gas_price, gas_limit, to, value, input, recovery_id, r, s)
    |> ExRLP.encode()
    |> Base.encode16(case: :lower)
  end

  @doc """
  Construct a transaction object in the Ethereum format, which is a list of nince elements' binary values:
  1. `nonce`: The sender's nonce at the current block. Evaluated as empty binary array (<<>>) if it is zero.
  2. `gas_price`:
  3. `gas_limit`:
  4. `to`: Receiver's address. Evaluated as empty binary array (<<>>) if it is null.
  5. `value`: The amount of Wei to transact. Evaluated as empty binary array (<<>>) if it is null or zero.
  6. `input`: The input data. Evaluated as empty binary array (<<>>) if it is null.
  7. `v`: Recover Id. Use chain_id to get a transaction data to sign; use either 37 or 38 (EIP-155) to get the raw transaction to broadcast.
  8. `r`: r value of the ECDS signature. Evaluated as empty binary array (<<>>) if it is null or zero.
  9. `v`: v value of the ECDS signature. Evaluated as empty binary array (<<>>) if it is null or zero.
  """
  @spec new_transaction(
          integer(),
          integer(),
          integer(),
          String.t(),
          integer(),
          String.t(),
          integer(),
          integer(),
          integer()
        ) :: list(binary())
  def new_transaction(nonce, gas_price, gas_limit, to, value, input, v, r, s) do
    to = to || ""
    value = value || nil
    input = input || ""

    [
      to_bytes(nonce),
      to_bytes(gas_price),
      to_bytes(gas_limit),
      Utils.hex_to_binary(to),
      to_bytes(value),
      Utils.hex_to_binary(input),
      to_bytes(v),
      to_bytes(r),
      to_bytes(s)
    ]
  end

  @doc """
  Generates the transaction hash. In Ethereum the data to be hash is the raw transaction.
  """
  @spec get_transaction_hash(String.t()) :: String.t()
  def get_transaction_hash(raw_transaction) do
    raw_transaction
    |> Utils.hex_to_binary()
    |> Keccak.kec()
    |> Base.encode16()
  end

  @doc """
  Gets the corresponding Ethereum address from the `private_key`.
  """
  @spec get_address_from_private_key(String.t()) :: String.t()
  def get_address_from_private_key(private_key) do
    {:ok, <<0x04, public_key::binary>>} =
      private_key |> Utils.hex_to_binary() |> Signature.get_public_key()

    <<_::binary-size(12), address::binary-size(20)>> = Keccak.kec(public_key)

    "0x" <> Base.encode16(address, case: :lower)
  end

  defp get_recovery_id(v) when v in [0, 1], do: Utils.get_chain_id() * 2 + 35 + v

  defp to_bytes(0), do: ""
  defp to_bytes(""), do: ""
  defp to_bytes(nil), do: ""

  defp to_bytes(number) when is_integer(number) do
    hex = Integer.to_string(number, 16)

    case rem(String.length(hex), 2) do
      0 -> Base.decode16!(hex)
      1 -> Base.decode16!("0" <> hex)
    end
  end
end


lib/eth/advance/transaction.ex
defmodule OcapRpc.Internal.EthTransaction do
  @moduledoc """
  Contains methods to construct and sign Ethereum transactions.
  """
  alias OcapRpc.Eth.Account
  alias OcapRpc.Eth.Chain
  alias OcapRpc.Eth.Transaction
  alias OcapRpc.Internal.EthTransaction.Helper

  @doc """
  Composes a transaction to sign.
  """
  def get_transaction_to_sign(from, to, value, opts \\ []) do
    nonce = get_nonce_by_address(from, opts)
    gas_price = get_gas_price(opts)
    gas_limit = get_gas_limit(opts)
    input = Keyword.get(opts, :input, nil)
    Helper.get_transaction_to_sign(nonce, gas_price, gas_limit, to, value, input)
  end

  @doc """
  Compose and sign a transaction, returns the raw transaction.
  """
  def compose_transaction(private_key, to, value, opts \\ []) do
    nonce = get_nonce_by_sk(private_key, opts)
    gas_price = get_gas_price(opts)
    gas_limit = get_gas_limit(opts)
    input = Keyword.get(opts, :input, nil)

    get_raw_transaction(nonce, gas_price, gas_limit, to, value, input, private_key)
  end

  @doc """
  Signs and then broadcasts the transaction, returns the transaction hash.
  """
  def send_transaction(private_key, to, value, opts \\ []) do
    private_key
    |> compose_transaction(to, value, opts)
    |> Transaction.send_raw()
  end

  defp get_raw_transaction(nonce, gas_price, gas_limit, to, value, input, private_key) do
    {_signature, r, s, v} =
      nonce
      |> Helper.get_transaction_digest_to_sign(gas_price, gas_limit, to, value, input)
      |> Helper.get_signature(private_key)

    "0x" <> Helper.get_raw_transaction(nonce, gas_price, gas_limit, to, value, input, v, r, s)
  end

  defp get_nonce_by_sk(private_key, opts) do
    private_key
    |> Helper.get_address_from_private_key()
    |> get_nonce_by_address(opts)
  end

  defp get_nonce_by_address(address, opts) do
    case Keyword.get(opts, :nonce, nil) do
      nil -> Transaction.get_transaction_count(address)
      nonce -> nonce
    end
  end

  defp get_gas_price(opts) do
    case Keyword.get(opts, :gas_price, nil) do
      nil -> Chain.gas_price()
      0 -> Chain.gas_price()
      gas_price -> gas_price
    end
  end

  defp get_gas_limit(opts) do
    gas_limit = Keyword.get(opts, :gas_limit, nil)
    input = Keyword.get(opts, :input, nil)
    get_gas_limit(gas_limit, input)
  end

  defp get_gas_limit(nil, input) when input in [nil, ""], do: 22_000
  defp get_gas_limit(nil, _), do: 30_000
  defp get_gas_limit(gas_limit, _), do: gas_limit
end


lib/ocap_rpc/application.ex
defmodule OcapRpc.Application do
  @moduledoc false

  use Application

  def start(_type, _args) do
    children = []
    opts = [strategy: :one_for_one, name: OcapRpc.Supervisor]

    update_config()

    Supervisor.start_link(children, opts)
  end

  defp update_config do
    app = :ocap_rpc
    update_conn(app, :btc)
    update_conn(app, :eth)
    update_conn(app, :ipfs)
    update_conn(app, :cmt)
  end

  defp update_conn(app, :btc = type) do
    conf = Application.get_env(app, type, [])

    conn =
      conf
      |> Access.get(:conn, %{port: 8332})
      |> Map.put(:hostname, System.get_env("BTC_RPC_HOST") || "localhost")
      |> Map.put(:user, System.get_env("BTC_RPC_USER"))
      |> Map.put(:password, System.get_env("BTC_RPC_PASS"))

    Application.put_env(app, type, Keyword.put(conf, :conn, conn))
  end

  defp update_conn(app, :eth = type) do
    conf = Application.get_env(app, type, [])

    conn =
      conf
      |> Access.get(:conn, %{port: 8545})
      |> Map.put(:hostname, System.get_env("ETH_RPC_HOST") || "localhost")

    Application.put_env(app, type, Keyword.put(conf, :conn, conn))
  end

  defp update_conn(app, :ipfs = type) do
    conf = Application.get_env(app, type, [])

    conn =
      conf
      |> Access.get(:conn, %{ort: 5001})
      |> Map.put(:hostname, System.get_env("IPFS_RPC_HOST") || "localhost")

    Application.put_env(app, type, Keyword.put(conf, :conn, conn))
  end

  defp update_conn(app, :cmt = type) do
    conf = Application.get_env(app, type, [])

    conn =
      conf
      |> Access.get(:conn, %{port: 8545})
      |> Map.put(:hostname, System.get_env("CMT_RPC_HOST") || "localhost")

    Application.put_env(app, type, Keyword.put(conf, :conn, conn))
  end
end
